lib/ddtrace/profiling.rb

```diff
@@ -4,24 +4,61 @@ module Profiling
     module_function
 
     GOOGLE_PROTOBUF_MINIMUM_VERSION = Gem::Version.new('3.0')
+    private_constant :GOOGLE_PROTOBUF_MINIMUM_VERSION
 
     def supported?
-      google_protobuf_supported?
+      unsupported_reason.nil?
     end
 
-    def native_cpu_time_supported?
-      require 'ddtrace/profiling/ext/cpu'
-      Ext::CPU.supported?
+    def unsupported_reason
```

Comment:
I really like having a string "reason" to explain why it isn't supported.

On the following if statements, would it make sense to break the conditionals into smaller, programmatic functions? Might help checking specific conditions. In this way we could preserve `google_protobuf_supported?`, `native_cpu_time_supported?` etc...

Also I might suggest that there are many "reasons" why it might be unsupported. Perhaps returning an Array instead of only the first would be better descriptive and allow users to fix multiple problems at once instead of fixing one, trying  again, fixing another until it works.

Reply:
> On the following if statements, would it make sense to break the conditionals into smaller, programmatic functions? Might help checking specific conditions. In this way we could preserve google_protobuf_supported?, native_cpu_time_supported? etc...

Would you see those individual methods as being public? 

If they'd be private, I'd go YAGNI for now -- all of the checks are nice one-liners, so I think it's quite manageable/readable to maintain them right now, and we can at any point revisit this decision.

If they'd be public, I'm not quite sure what the usefulness of that would be, and it'd be slightly constraining for evolution, e.g. if we find a cpu time approach that works for all rubies OR if we move to an alternative protobuf implementation that just works, we'd keep these weird appendix methods around that don't seem to offer too much value.

> Also I might suggest that there are many "reasons" why it might be unsupported. Perhaps returning an Array instead of only the first would be better descriptive and allow users to fix multiple problems at once instead of fixing one, trying again, fixing another until it works.

Yess....ish. Doing that would complicate the checks/mechanism because right now every check can implicitly assume that all checks above have cleared, and gathering up all the reasons would kind-of break this. For instance, there's three protobuf checks: C1 - is it here, C2 -  is it old, C3 - and can we load it. C2 only makes sense if C1 passes; and C3 only makes sense if C2 passes.

This is of course solveable -- we could refactor this to take these dependencies into account explicitly, but... I'm not quite sure it's worth it, honestly.

It's quite a bit of work, quite a bit more complexity, and in the end, I'm guessing it would benefit very few customers -- I'm not sure the unlucky person that has a JRuby 1.7 app, than then tries to move it to Ruby 2.0, forgets to add google-protobuf, etc exists :)

Reply:
> Would you see those individual methods as being public?

Yes. We've had users call them to debug specific issues before. They're useful and I don't see any extra cost in exposing them, as they were before. It also helps make the unit test logic much simpler (reduced branch complexity.)

> It's quite a bit of work, quite a bit more complexity, and in the end, I'm guessing it would benefit very few customers

I'm not of that opinion; when helping users get setup I found that there was often more than one issue, and that it was frustrating thinking it had been solved, when it turned out there was more to do. This was especially annoying when not done locally, but through deploys to some external environment where the feedback loop was slower and each restart took more time.

Reply:
> I'm not of that opinion; when helping users get setup I found that there was often more than one issue, and that it was frustrating thinking it had been solved, when it turned out there was more to do. This was especially annoying when not done locally, but through deploys to some external environment where the feedback loop was slower and each restart took more time.

Hmm. I don't agree -- I think for the checks we do right now, it's a really corner case to hit them, so I'd rather spend the extra effort elsewhere. Do you see this change as blocking this PR?

Reply:
Perhaps the "multiple reasons" are not blocking, since it's an expansion on the existing functionality.

But I do consider the individual methods as blocking, as it was a public API, would effectively reduce the branch complexity of `unsupported reason`, and it takes little effort to do.

Reply:
Expanding on my reasoning: I _really_ don't want to make internal design constraints into public APIs that we'll have to maintain.

`google_protobuf_supported?` is only a problem if we're relying on an external implementation of protobuf, as we do currently.

If we stop using protobuf as a format _or_ if we embed a working implementation of protobuf _or_ if we move the probobuf encoding to the agent _or_ some other option; then this method would not make any sense anymore.
Making it a public API would force us to maintain it. (Would we use `true` or `false` when there's no protobuf, but we don't need it anyway?)

I see the same thing as applying to other constraints -- they may change over time. We may add support for jruby; we may re-add support for older rubies; we may add support for protobuf 2.x, etc. None of them is set in stone.

I don't like it, but I can live with breaking `unsupported_reason` into parts. BUT not with breaking it AND making the components into public APIs.

Since it seems we're at a stand-still I think we need a tie-breaker. My suggestion is: Would you mind picking someone to add a second opinion here? I commit to accepting what the second opinion decides.

Reply:
I think that a clear signal that it failed and a reason message are enough.
With such information, the client is able to assess what do next: either address the issue reported and try again, or accept that their application is not compatible with the profiler.

Reply:
Regardless of whether certain functions become public or not, the issue remains that the following adds high branch complexity:

```ruby
if RUBY_ENGINE == 'jruby'
elsif Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.1')
elsif Gem.loaded_specs['google-protobuf'].nil?
elsif Gem.loaded_specs['google-protobuf'].version < GOOGLE_PROTOBUF_MINIMUM_VERSION
elsif !protobuf_loaded_successfully?
end
```

Each one of these scenarios being dependent on the other in a specific order makes it naturally more difficult to exhaustively test this. My suggestion to at least move the conditions after the `if/elsif` to their own functions e.g. `jruby?` or `google_protobuf_available?` would make the code easier to read, test, and you could still mark them `private` if you're concerned about adding things to the public API you don't want to expose.

Reply:
Fixed in 01c0f72d880ea27fdf73f371628228b1904965f1